### PR TITLE
Improve gates: add multiple gates as infrastructure, for more convenience algorithm test

### PR DESCRIPTION
**Pull request by henankf223**
- [x] Add V and cV gates (sqrt-NOT, V^2 = X)
- [x] Add a general control-U builder: make_control_gate(size_t control, QuantumGate& U)
- [x] Add advance_gate_helper.py, which implement 3-qubit gates as an efficient decomposition circuit:
    Toffoli: Control-Control-NOT gate
    Fredkin: Control-SWAP gate

- [x] Adjust affected scripts and test cases (qft)
- [x] Test cases for new gates: advanced_gates_test.py
- [x] Ready to go

### CODE REVIEW

File: src/qforte/helper/advance_gates_helper.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    F_circ.add_gate(C32)
+    F_circ.add_gate(CV23)
+    F_circ.add_gate(CV13)
+    F_circ.add_gate(C12)
+    F_circ.add_gate(CV23.adjoint())
+    F_circ.add_gate(C32)
+    F_circ.add_gate(C12)
+
+    return F_circ
+
```
Comment (fevangelista) [inline src/qforte/helper/advance_gates_helper.py]: There seem to be about 10 empty lines here, can you please remove them?

File: src/qforte/utils/qft.py [hunk 2]
```diff
@@ -3,32 +3,34 @@
 
 
 
-def qft_circuit(n, direct):
+def qft_circuit(na, nb, direct):
```
Comment (fevangelista) [inline src/qforte/utils/qft.py]: Can you please make sure that the function is well-behaved if `direct` is ill specified (e.g. due to a typo?). The right thing to do would be to raise an exception.

File: src/qforte/helper/advance_gates_helper.py (hunk 1 above)
Comment (henankf223) [reply src/qforte/helper/advance_gates_helper.py]: Sure, I have commited the changes that remove empty lines

File: src/qforte/utils/qft.py (hunk 2 above)
Comment (henankf223) [reply src/qforte/utils/qft.py]: Sure, I have added an exception here, thanks for the advice!

File: src/qforte/utils/qft.py [hunk 3]
```diff
@@ -3,55 +3,60 @@
...
+            qft_circ.add_gate(qforte.make_gate('SWAP', i+na, lens-1-i+na))
 
     if direct == 'forward':
         return qft_circ
-    if direct == 'reverse':
+    elif direct == 'reverse':
         return qft_circ.adjoint()
-
+    else:
+        raise ValueError('QFT directions can only be "forward" or "reverse"')
```
Comment (fevangelista) [inline src/qforte/utils/qft.py]: Anytime we catch a wrong value we should raise a `ValueError`.

File: src/qforte/utils/qft.py [hunk 4]
```diff
@@ -3,55 +3,60 @@
...
     else:
-        for i in range(int((n-1)/2)):
-            qft_circ.add_gate(qforte.make_gate('SWAP', i, n-1-i))
+        for i in range(int((lens-1)/2)):
+            qft_circ.add_gate(qforte.make_gate('SWAP', i+na, lens-1-i+na))
 
     if direct == 'forward':
         return qft_circ
-    if direct == 'reverse':
+    elif direct == 'reverse':
```
Comment (fevangelista) [inline src/qforte/utils/qft.py]: I simplified the logic a bit here.

**Comment by fevangelista** [comment]
Ok, I have made two minor changes. When the tests pass let's merge this.

**Comment by henankf223** [comment]
Thanks!

**Pull request merged**